File: src/main/java/example/domain/model/specification/TrainType.java
package example.domain.model.specification;

import example.domain.model.bill.Amount;

/**
 * 列車種類
 */
public enum TrainType {
    のぞみ(true),
    ひかり(false);

    private boolean isApply;
    private TrainType(boolean isApply) {
        this.isApply = isApply;
    }

    public Amount apply(Amount surcharge) {
        return isApply ? surcharge : new Amount(0);
    }

    public static TrainType valueOf(Amount surcharge) {
        return surcharge.equals(new Amount(0)) ? ひかり : のぞみ;
    }
}

File: jr-pricing/Model/Specification/TrainType.vb
Imports System.Runtime.CompilerServices
Imports Jr.Pricing.Model.Bill

Namespace Model.Specification
    ''' <summary>
    ''' 列車種類
    ''' </summary>
    ''' <remarks></remarks>
    Public Enum TrainType
        のぞみ
        ひかり
    End Enum
    Public Module TrainTypeExtension
        Private ReadOnly ひかりSurcharge As New Amount(0)
        ''' <summary>
        ''' 振る舞いを返す
        ''' </summary>
        ''' <param name="aType">区分</param>
        ''' <returns>振る舞い</returns>
        ''' <remarks></remarks>
        <Extension()> Public Function Calculate(aType As TrainType, additionalSurcharge As Amount) As Amount
            Return If(aType = TrainType.ひかり, ひかりSurcharge, additionalSurcharge)
        End Function

        <Extension()> Public Function ValueOf(aType As TrainType, additionalSurcharge As Amount) As TrainType
            Return If(ひかりSurcharge.Equals(additionalSurcharge), TrainType.ひかり, TrainType.のぞみ)
        End Function
    End Module
End Namespace